Bindings/Java/Matlab/Hopper_Device/OptimizeHopper.m

```diff
@@ -25,81 +26,184 @@
 % permissions and limitations under the License.                        %
 %-----------------------------------------------------------------------%
 
-% Three variables: 1) passive stiffness 
-%                  2) active max torque 
-%                  3) hop timing
-% See buildHopperFromSolution() for more details.
-numVariables = 3; 
+% CHOOSE PROBLEM: 'one_hop' or 'three_hops'
+problem = 'one_hop';
+switch problem
+    
+    % Optimize the height of one hop. The single optimization variable 
+    % determines when to simultaneously release a loaded passive device and 
+    % activate the hopper vastus muscle (see CONSTRUCTCONTROLS below). This
+    % simple univariate problem uses FMINBND from the MATLAB Optimization 
+    % Toolbox. 
+    case 'one_hop'
+        lb = 0;
+        ub = 1;         
+        options = optimset('TolFun', 1e-2, ...
+                           'MaxIter', 10000, ...
+                           'Display', 'iter'); 
+        [x, f] = fminbnd(@(x) objective(x,problem),lb,ub,options);
+                     
+    % Optimize the height of a sequence of three hops. The optimization
+    % variables decide to switch the actuators on and off to load and 
+    % release the passive device with the activated muscle (see 
+    % CONSTRUCTCONTROLS below). This problem is subject to falling into
+    % undesired local minima (e.g. solutions with less than three hops) and
+    % requires an optimizer from the MATLAB Global Optimization Toolbox.
+    % Here we use the genetic algorithm optimizer, GA.
+    case 'three_hops'
+        numVars = 3; 
+        lb = 0.25*ones(numVars, 1);
+        ub = 0.75*ones(numVars, 1);
+        options = optimoptions('ga','FunctionTolerance', 1e-2, ...
+                                    'MaxGenerations', 10, ...
+                                    'Display', 'iter'); 
+        [x, f] = ga(@(x) objective(x,problem),numVars,[],[],[],[], ...
+                                   lb,ub,[],options);                 
+end
+
+% ANALYZE OPTIMIZED SOLUTION
+% Construct the actuator controls from solution 'x' and interpolate. See 
+% CONSTRUCTCONTROLS below for details.
+[deviceControl, muscleExcitation] = constructControls(x, problem);
+time = 0:0.01:5;
+deviceControlInterp    = interp1(deviceControl(1,:), ... 
+                                 deviceControl(2,:), time);
+muscleExcitationInterp = interp1(muscleExcitation(1,:), ... 
+                                 muscleExcitation(2,:), time);
```

Comment:
Is the interpolation necessary? The control is piecewise linear.

Reply:
No it's not. Not sure why I included it...